prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -1130,160 +1138,262 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+                    operand,
+                    times,
+                    ty,
+                    location,
+                )?
+            }
+            ref rhs => {
+                unimplemented!("encoding of '{:?}'", rhs);
+            }
+        })
     }
 
     /// Encode the lhs and the rhs of the assignment that create the loan
     fn encode_loan_places(&mut self, loan_places: &LoanPlaces<'tcx>) -> EncodingResult<(vir::Expr, vir::Expr, bool, Vec<vir::Stmt>)> {
         debug!("encode_loan_places '{:?}'", loan_places);
+
         // will panic if attempting to encode unsupported type
-        let (expiring_base, mut stmts, expiring_ty, _) = self.encode_place(&loan_places.dest).unwrap();
+        let (expiring_base, mut stmts, expiring_ty, _) = self.encode_place(&loan_places.dest, ArrayAccessKind::Mutable(None, loan_places.location)).unwrap();
+        trace!("expiring_base: {:?}", (&expiring_base, &stmts, &expiring_ty));
+
+        // the original encoding of arrays is with a sort-of magic temporary variable, so
+        // `postprocess_place_encoding` will return `i32` here instead of Array$3$i32. so here
+        // we call the mir_encoder one that returns a `PlaceEncoding`, and do the
+        // postprocessing if necessary ourselves.
+        if let mir::Rvalue::Ref(_, mir::BorrowKind::Mut { .. }, ref rhs_place) = loan_places.source {
+            let (rhs_place_encoding, ..) = self.mir_encoder.encode_place(rhs_place).unwrap();
+            if let PlaceEncoding::ArrayAccess { .. } = rhs_place_encoding {
+                // encode expiry of the array borrow
+                let (expired_expr, regained_array, wand_rhs) = self.array_magic_wand_at[&loan_places.location].clone();
+
+                // expiring base is something like ref$i32, so we need .val_ref.val_int
+                let deref = self.encoder.encode_value_expr(expiring_base.clone(), expiring_ty);
+                let target_ty = if let ty::TyKind::Ref(_, target_ty, _) = expiring_ty.kind() { target_ty } else { unreachable!() };
+                let expiring_base_value = self.encoder.encode_value_expr(deref, target_ty);
+
+                // the original magic wand refered to the temporary variable that we created for array
+                // encoding. the expiry here refers to the non-temporary rust variable, so we need
+                // to replace that in the wand RHS
+                let wand_rhs = wand_rhs.replace_place(&expired_expr, &expiring_base_value);
+
+                // we can't refer to the somewhat magic "lhs" label here, because there is no wand,
+                // so there's no LHS. instead, we add a label just before the inhale, and refer to
+                // that
+                let new_lhs_label = self.cfg_method.get_fresh_label_name();
+                stmts.push(
+                    vir::Stmt::Label(new_lhs_label.clone())
+                );
+
+                let wand_rhs_patched_lhs = wand_rhs.fold_expr(
```

Comment:
Isn't there already a utility function to replace a particular old label somewhere in `vir`? I remember using a similar method in the (still unmerged) closures PR, but I'm not sure if I added it.

Reply:
indeed, found `Expr::map_old_expr_label`, thanks :-)